envoy/matcher/matcher.h

```diff
@@ -125,75 +126,79 @@ template <class DataType> class OnMatchFactory {
+  bool isComplete() const { return !isInsufficientData(); }
+  bool isNoMatch() const { return absl::holds_alternative<NoMatch>(result_); }
+  bool isMatch() const { return absl::holds_alternative<ActionFactoryCb>(result_); }
+  ActionFactoryCb actionFactory() const { return absl::get<ActionFactoryCb>(result_); }
+  ActionPtr action() const { return actionFactory()(); }
 };
 
 // Callback to execute against skipped matches' actions.
-template <class DataType> using SkippedMatchCb = std::function<void(const OnMatch<DataType>&)>;
+using SkippedMatchCb = std::function<void(ActionFactoryCb)>;
 /**
  * MatchTree provides the interface for performing matches against the data provided by DataType.
  */
 template <class DataType> class MatchTree {
 public:
   virtual ~MatchTree() = default;
 
-  // The result of a match. There are three possible results:
-  // - The match could not be completed (match_state_ == MatchState::UnableToMatch)
-  // - The match was completed, no match found (match_state_ == MatchState::MatchComplete, on_match_
-  // = {})
-  // - The match was complete, match found (match_state_ == MatchState::MatchComplete, on_match_ =
-  // something).
-  struct MatchResult {
-    const MatchState match_state_;
-    const absl::optional<OnMatch<DataType>> on_match_;
-  };
-
   // Attempts to match against the matching data (which should contain all the data requested via
-  // matching requirements). If the match couldn't be completed, {false, {}} will be returned.
-  // If a match result was determined, {true, action} will be returned. If a match result was
-  // determined to be no match, {true, {}} will be returned.
+  // matching requirements).
+  // If the match couldn't be completed, MatchResult::insufficientData() will be returned.
+  // If a match result was determined, an action callback factory will be returned.
+  // If it was determined to be no match, MatchResult::noMatch() will be returned.
+  //
+  // Implementors should call handleRecursionAndSkips() to transform OnMatch values
+  // into MatchResult values, and handle noMatch and insufficientData results as appropriate
+  // for the specific matcher type.
   virtual MatchResult match(const DataType& matching_data,
-                            SkippedMatchCb<DataType> skipped_match_cb = nullptr) PURE;
+                            SkippedMatchCb skipped_match_cb = nullptr) PURE;
 
 protected:
   // Internally handle recursion & keep_matching logic in matcher implementations.
   // This should be called against initial matching & on-no-match results.
   static inline MatchResult
   handleRecursionAndSkips(const absl::optional<OnMatch<DataType>>& on_match, const DataType& data,
-                          SkippedMatchCb<DataType> skipped_match_cb) {
+                          SkippedMatchCb skipped_match_cb) {
     if (!on_match.has_value()) {
-      return {MatchState::MatchComplete, absl::nullopt};
+      return MatchResult::noMatch();
     }
     if (on_match->matcher_) {
```

Comment:
on_match could use a variant! then a switch or a visit would make it easier to parse.

Reply:
Yes, but doesn't have to be in this PR. (And that one isn't as big of an improvement since it doesn't get rid of the template. I think it's *possible* to get rid of the template too, but that's a much more aggressive change.)